### PR TITLE
Initial nixified build of this binary

### PR DESCRIPTION
**Pull request by Ocramius**
This approach relies on @Hoverbear's excellent guide at https://github.com/Hoverbear/hoverbear.org/blob/417d50e1050cf201c88e74681074803a86ccd018/content/blog/2021-06-25-a-flake-for-your-crate/index.md

Ref: Hoverbear/hoverbear.org#a9abd75063a7eed7bccf9d7918fe1f8561a20c8a

With this change, we can run:

```sh
nix develop
nix flake check
nix build .
nix run .
```

After this, I'll pursue:

 * cleanup
 * dockerization of this built result (put it in a scratch container)

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -8,17 +8,8 @@ jobs:
...
-          push: "false"
-          load: "true"
-          cache-from: "type=gha,scope=ci-cache"
-          cache-to: "type=gha,mode=max,scope=ci-cache"
-
+      - uses: actions/checkout@v4
+      - uses: DeterminateSystems/nix-installer-action@v4
+      - uses: DeterminateSystems/magic-nix-cache-action@v2
+      - run: nix flake check
+      - run: nix build .\#haha
```
Comment (Ocramius) [inline .github/workflows/ci.yml]: Name to be adjusted

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,86 @@
+# https://github.com/Hoverbear/hoverbear.org/blob/417d50e1050cf201c88e74681074803a86ccd018/content/blog/2021-06-25-a-flake-for-your-crate/index.md#flakenix
+{
+  description = "My cute Rust crate!";
```
Comment (Ocramius) [inline flake.nix]: This obviously requires `README.md` and `Makefile` to be adjusted too

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+              Env = [
+                "OCI_SRM_SERVER_MOCK_PORT=80"
+                "OCI_SRM_SERVER_MOCK_BASE_URL=http://oci-srm-server-mock/"
+                "PUNCHOUT_SERVER_LOGIN_URI=http://punchout-server/punch-in?foo=bar&pass=example-supersecret"
+                "PUNCHOUT_SERVER_CONFIRMATION_URI=http://punchout-server/cxml-order-request-endpoint"
+              ];
+              ExposedPorts = { "80/tcp" = { }; };
+            };
+          };
+        });
```
Comment (Ocramius) [inline flake.nix]: `Dockerfile` can be removed too

File: flake.nix (hunk 2 above)
Comment (Ocramius) [reply flake.nix]: text to be adjusted here too

File: Dockerfile [hunk 4]
```diff
@@ -1,3 +1,4 @@
+# @TODO this file is to be removed: it only exists for reference, until we can get static compilation to work in Nix too
```
Comment (Ocramius) [inline Dockerfile]: To be fixed before merging, IMO, but my nix-fu is too weak today, and I need to figure out how to get `-musl` builds running :)

**Comment by Ocramius** [comment]
@ciaranmcnulty just an example, FYI: I removed `docker` from the build-side, and only use it for `load`, `push` and `tag` here, with fully locked dependencies (down to the tiniest detail).

Hope this is useful on your end too :)

File: Dockerfile (hunk 4 above)
Comment (Ocramius) [reply Dockerfile]: Ref: https://old.reddit.com/r/NixOS/comments/f0yi3b/how_to_build_a_simple_static_rust_binary_using/

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            {
+              buildInputs = with pkgs; [ rustfmt cargo ];
+            } ''
+            #${pkgs.rustfmt}/bin/cargo-fmt fmt --manifest-path ${./.}/Cargo.toml -- --check
+            ${pkgs.nixpkgs-fmt}/bin/nixpkgs-fmt --check ${./.}
+            touch $out # it worked!
+          '';
+          "${cargoToml.package.name}" = pkgs."${cargoToml.package.name}";
+        });
+      devShell = forAllSystems (system:
```
Comment (Hoverbear) [inline flake.nix]: You may get a warning on more recent nixes, I believe convention has drifted to `devShells.default` (eg https://github.com/DeterminateSystems/nix-installer/blob/8395b40223f56a12f97f29750b6fc8feea80ecb1/flake.nix#L123-L159)

Comment (Ocramius) [reply flake.nix]: Thanks for poking me about this: agreed, I'm indeed simplifying a lot of the stuff until I get only the bits I understand out of it. I'm indeed getting a warning, which I'll chase later :+1: 

The final result I'll get may not be abstracted, or very reusable, but my current aim is just getting the docker image out of this, for now :D

Comment (Ocramius) [reply flake.nix]: 8af03669e66cbebcd631c835ddca6a4cbc82b846

Everything is a bit more rigid, and probably also wrong, but works for now :D

**Comment by Ocramius** [comment]
```
warning: The check omitted these incompatible systems: aarch64-darwin, aarch64-linux, x86_64-darwin
Use '--all-systems' to check all.
```

This is fine

:shipit: 

Also, @Hoverbear, amazing job with the magic cache! :cake:

**Pull request merged**